fix: ensure cypress in cypress uses prod dependencies from the snapshot

ryanthemanuel opened the pull request:
### User facing changelog
<!-- 
Explain the change(s) for every user to read in our changelog. Examples: https://on.cypress.io/changelog
If the change is not user-facing, write "n/a".
-->

n/a

### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

Due to the way v8 snapshots work, we need to ensure that the dependencies are loaded via an entry point. Since the e2ePluginSetup code is sort of piecemeal the initial imports/requires are not loaded from the snapshot. By having every prod dependency go through an entry point, we will ensure that everything is loaded consistently from the snapshot. We only should do this with prod dependencies.

### Steps to test
<!--
For non-trivial behavior changes, list the steps that a reviewer should follow to validate the new behavior.
This is not meant to be the only testing performed by a reviewer, just the "happy path" that leads to the new behavior.
-->

n/a

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->

n/a

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [na] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

ryanthemanuel committed:
fix: ensure cypress in cypress uses prod dependencies from the snapshot - run ci

ryanthemanuel committed:
fix: ensure cypress in cypress uses prod dependencies from the snapshot - run ci

ryanthemanuel committed:
refactoring

ryanthemanuel committed:
refactoring

ryanthemanuel committed:
refactoring -- run ci

lmiller1990 commented:
This didn't cause any problems until @astone123 added some new deps in a feature branch. Why did adding those dependencies cause this to suddenly become an issue?

Also, any tips to debug this? How did you 

1) figure this out
2) land on this specific fix

lmiller1990 commented on tooling/v8-snapshot/cache/darwin/snapshot-meta.json:
```diff
@@ -871,7 +871,7 @@
     "./packages/server/lib/util/glob.js",
     "./packages/server/lib/util/proxy.ts",
-    "./packages/server/lib/util/resourceTypeAndCredentialManager.ts",
+    "./packages/server/lib/util/requestedWithAndCredentialManager.ts",
```
How did this get added? Was it automatic? If not, how did you decide to add it?

ryanthemanuel replied:
This would normally be generated by my auto generator script so normally we shouldn't see these on individual pull requests, but I explicitly made this change locally just to make sure things were working in "prod mode" on my machine.

lmiller1990 reviewed (approved):
Thanks for this - questions about how you debug this!

ryanthemanuel commented:
> This didn't cause any problems until @astone123 added some new deps in a feature branch. Why did adding those dependencies cause this to suddenly become an issue?
> 
> Also, any tips to debug this? How did you
> 
> 1. figure this out
> 2. land on this specific fix

I think it was ultimately a timing problem. The problem was still occurring but we were never noticing it. This problem manifested on @astone123's branch after I fixed the babel dependency problem.

For the babel dependency problem:

* It didn't have a chance to manifest itself until after we added react-docgen which I believe was the first point where a fair amount of babel dependencies were being added to the snapshot. Previously, babel was only used in the child process (which doesn't use the snapshot and shouldn't in prod)
* I think this actually is an example of cypress in cypress and snapshots exposing problems. The ultimate root cause was that we were deferring some code that properly populates the [`TYPES`](https://github.com/babel/babel/blob/v7.17.7/packages/babel-traverse/src/path/index.ts#L253) object in babel. Thus, the TYPES object was not properly populated.
* By norewriting [like we did](https://github.com/cypress-io/cypress/commit/f054bdcb14a0cc3c368a7debdbb05f9b86529d5e) this allows for this code to be loaded in time.
* To troubleshoot this problem, based on the error I was seeing I reasoned it had to be with some of the babel dependencies. So I tried moving all of them to norewrite and it fixed the problem, and then I whittled it down to where the problem was and reasoned about what was going on.

For the problem fixed in this PR:

* This one was trickier. This problem was manifesting itself as srv being undefined in [this](https://github.com/cypress-io/cypress/blob/4726ea90a9acdb6dec75616be29f5b761f17ecb8/packages/graphql/src/makeGraphQLServer.ts#L94) line of code. 
* I added some debug logging and determined that `makeGraphQLServer` was being imported twice which clued me into it wasn't actually properly being loaded in the snapshot (and then being loaded again from another dependency within the snapshot).

mjhenkes reviewed (approved):
Better comment plz

ryanthemanuel committed:
added a clearer comment

lmiller1990 commented:
Thanks for the explanation @ryanthemanuel :+1:
